feat(certificatemanager): requesting private certificates issued by Private Certificate Authority

jumic opened the pull request:
Support requesting private certificates issued by Private Certificate Authority. 

Similar to the existing construct named `Certificate`, a new construct `PrivateCertificate` was introduced. There are two main differences between them. `PrivateCertificate` has an additional property `certificateAuthority` to specify the Private certificate authority (CA) that will be used to issue the certificate. The validation options are removed because no validation is necessary for private certificates.

Closes #10076.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: a0746343fe977f40f924d194b45ae5caa4c0eb75
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4a1871b6-0b0d-4610-84d4-db3ebdbefdda%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

njlynch reviewed (dismissed):
Thanks for this PR!

I'm wondering if we should go about this differently. Rather than add this to `Certificate`, what if we create a new construct called `PrivateCertificate`? The `PrivateCertificateProps` would be limited to `domainName`, `subjectAlternativeNames`, and `certificateAuthority`, with the latter required rather than optional.

I think this will make the intent a bit clearer. There's a public certificate -- with various validation options -- and a private certificate, which requires the use of a private CA and requires no validation. I suspect this will result in less build/synth-time errors and slightly cleaner code.

jumic commented:
Thanks for your feedback. I have implemented the suggested changes and introduced a new construct named `PrivateCertificate. You're right. The code looks much cleaner than before.

I tested the new construct manually with the following examples:

```ts
new certificatemanager.PrivateCertificate(this, 'MyCert', {
  certificateAuthority,
  domainName: 'my-cert.me',
});
```

```ts
new certificatemanager.PrivateCertificate(this, 'MyCert', {
  certificateAuthority,
  domainName: 'my-cert.me',
  subjectAlternativeNames: ['*.foo.my-cert.me'],
});
```

I haven't created an integration test yet because `PrivateCertificate` depends on `acmpca.CertificateAuthority` which has to be created manually before. Please let me know if I should add the integration test anyway.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: a9bf0c73f1aad22dd703d254eea3cfc678129468
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3dea0d50-27f0-4850-a061-fb446bbfcad8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: e39cc1cace294dea68c0e5da49b77794543e6b71
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=cf96f119-f8ca-4ff5-9014-0f2a3de31f4b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: d1634501311a7a3a3bb5a2a05fd7eed35de64080
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fe9b690c-c953-40b6-aaea-e3ade4c36827%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
